src/sentry/static/sentry/app/views/organizationIntegrationConfig.jsx

```diff
@@ -0,0 +1,305 @@
+  color: ${p => p.theme.gray2};
+`;
+
+const MetadataLink = styled.a`
+  display: block;
+  margin-bottom: 6px;
+`;
+
+const alertLinkMarkedRenderer = new marked.Renderer();
+alertLinkMarkedRenderer.paragraph = s => s;
+
+function computeCenteredWindow(width, height) {
+  const screenLeft = window.screenLeft != undefined ? window.screenLeft : screen.left;
+  const screenTop = window.screenTop != undefined ? window.screenTop : screen.top;
+  const innerWidth = window.innerWidth
+    ? window.innerWidth
+    : document.documentElement.clientWidth
+      ? document.documentElement.clientWidth
+      : screen.width;
+
+  const innerHeight = window.innerHeight
+    ? window.innerHeight
+    : document.documentElement.clientHeight
+      ? document.documentElement.clientHeight
+      : screen.height;
+
+  const left = innerWidth / 2 - width / 2 + screenLeft;
+  const top = innerHeight / 2 - height / 2 + screenTop;
+
+  return {left, top};
+}
+
+export default class OrganizationIntegrationConfig extends AsyncView {
+  componentDidMount() {
+    this.dialog = null;
+    window.addEventListener('message', this.receiveMessage, false);
+  }
+
+  componentWillUnmount() {
+    super.componentWillUnmount();
+    window.removeEventListener('message', this.receiveMessage);
+
+    if (this.dialog !== null) {
+      this.dialog.close();
+    }
+  }
+
+  getProvider() {
+    const {config} = this.state;
+
+    if (config !== null) {
+      return config.providers.find(p => p.key == this.props.params.providerKey) || null;
+    }
+
+    return null;
+  }
+
+  getTitle() {
+    const provider = this.getProvider();
+
+    if (provider === null) {
+      return 'Global Integrations';
+    }
+
+    return `${provider.name} Integration`;
+  }
+
+  getEndpoints() {
+    const {orgId} = this.props.params;
+
+    return [
+      ['config', `/organizations/${orgId}/config/integrations/`],
+      ['itemList', `/organizations/${orgId}/integrations/`, {query: {status: ''}}],
+    ];
+  }
+
+  handleAddIntegration = provider => {
+    const name = 'sentryAddIntegration';
+
+    const {url, width, height} = provider.setupDialog;
+    const {left, top} = computeCenteredWindow(width, height);
+
+    this.dialog = window.open(
+      url,
+      name,
+      `scrollbars=yes,width=${width},height=${height},top=${top},left=${left}`
+    );
+
+    this.dialog.focus();
+    this.dialog.onclose = () => document.location.refresh();
+  };
+
+  receiveMessage = message => {
+    if (message.origin !== document.origin) {
+      return;
+    }
+
+    if (message.source !== this.dialog) {
+      return;
+    }
+
+    this.dialog = null;
+
+    const {success, data} = message.data;
+
+    if (!success) {
+      // TODO: Error state here
+      return;
+    }
+
+    // Merge the new integration into the list. If we're updating an
+    // integration ovewrrite the old integration.
+    const keyedItems = _.keyBy(this.state.itemList, 'id');
+    const itemList = sortArray(
```

Comment:
wouldn't it be simpler to keep it as an object in state and then turn it to a sorted array in render only? to avoid going back and forth?

Reply:
Nah, I'm ok with it working like this, Just have to have this merging logic since adding an integration is a create or update operation.